Fix buffering handling in progressive playback

filipe-norte opened the pull request:
When performing a seek operation during playback, the pipeline switches to the PAUSED state and starts buffering data. As this transition is async, we may reach the 100% buffer level before the state change completes. This may cause updateBufferingStatus() to stop getting called while the async transition is still in progress. As m_isBuffering was getting set back to m_wasBuffering value in updateStates() async state handling, once the state change completed, the updateStates() might not detect that we are not buffering anymore, as m_isBuffering is holding the wrong value.
<!--EWS-Status-Bubble-Start-->
https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/4fcc021a4d3e473b02453e2076c8b5d63a46f09f

| Build-Tests | Layout-Tests |
| ----------- | ------------ |
| | 
| | 
| [✅ 🛠 wpe-246-amd64-build](https://ews-wpe-rdk.igalia.com/#/builders/11/builds/166 "Built successfully") | [❌ 🧪 wpe-246-amd64-layout](https://ews-wpe-rdk.igalia.com/#/builders/9/builds/74 "Found 200 new test failures: compositing/geometry/clipped-video-controller.html, compositing/geometry/video-fixed-scrolling.html, compositing/geometry/video-opacity-overlay.html, compositing/shared-backing/clipping-and-shared-backing.html, compositing/video-page-visibility.html, compositing/video/video-reflection.html, compositing/visibility/visibility-simple-video-layer.html, imported/w3c/web-platform-tests/html/canvas/element/manual/imagebitmap/createImageBitmap-invalid-args.html, imported/w3c/web-platform-tests/html/canvas/element/manual/imagebitmap/createImageBitmap-serializable.html, imported/w3c/web-platform-tests/html/canvas/element/manual/imagebitmap/createImageBitmap-transfer.html ... (failure)") 
| [✅ 🛠 wpe-246-arm32-build](https://ews-wpe-rdk.igalia.com/#/builders/12/builds/167 "Built successfully") | [❌ 🧪 wpe-246-arm32-layout](https://ews-wpe-rdk.igalia.com/#/builders/10/builds/76 "Found 157 new test failures: compositing/geometry/clipped-video-controller.html, compositing/geometry/video-fixed-scrolling.html, compositing/geometry/video-opacity-overlay.html, compositing/overflow/scroll-ancestor-update.html, compositing/shared-backing/clipping-and-shared-backing.html, compositing/video-page-visibility.html, compositing/video/video-reflection.html, compositing/visibility/visibility-simple-video-layer.html, fullscreen/full-screen-iframe-legacy.html, imported/w3c/web-platform-tests/html/canvas/element/manual/imagebitmap/createImageBitmap-invalid-args.html ... (failure)") 
<!--EWS-Status-Bubble-End-->

filipe-norte-red committed:
Fix buffering handling in progressive playback

When performing a seek operation during playback, the pipeline switches
to the PAUSED state and starts buffering data. As this transition is
async, we may reach the 100% buffer level before the state change
completes. This may cause updateBufferingStatus() to stop getting
called while the async transition is still in progress.
As m_isBuffering was getting set back to m_wasBuffering value
in updateStates() async state handling, once the state change
completed, the updateStates() might not detect that we are not
buffering anymore, as m_isBuffering is holding the wrong value.

filipe-norte-red commented:
Thanks for the prompt reply, @eocanha. I appreciate the extra validation, as this was not a straightforward change considering all different execution paths.

Issue happens with playbin (gstreamer 1.18) and in-memory buffering (media disk cache is disabled). My use case is "(5) that buffering message might not come" .

You can find the logs in the attached [logs.zip](https://github.com/user-attachments/files/21620526/logs.zip) file.

The execution sequence is as follows:

- Pipeline is switched from PLAYING to PAUSED state
- Seek is triggered (timestamp 0:00:50.154692187)
- updateBufferingStatus() is called with "reset history" flag set.
    - This sets m_isBuffering to true and m_wasBuffering to true due "reset history" flag
- Async state change is in progress
- updateBufferingStatus() is called a few times with increasing buffering level, but reaches eventually 100%
    -  This sets m_isBuffering to false while m_wasBuffering is true (timestamp 0:00:51.206861128), BUT
    - updateStates() is called, which sets m_isBuffering to true again to delay true -> false transition (timestamp 0:00:51.206954836, and others)
- updateBufferingStatus() stops being called while we are still in the async state transition
- updateStates() is called with the state change to PAUSED now completed (timestamp 0:00:52.274080382), but m_isBuffering is still true and not reset afterward (no further calls to updateBufferingStatus()), so player believes we are still buffering

As on-disk buffering is disabled, the "MediaPlayerPrivateGStreamer::fillTimerFired()" won't get called to update the buffering status and recover from this situation.

Let me know if anything remains unclear.

Thanks

eocanha commented:
> updateBufferingStatus() stops being called while we are still in the async state transition

That's not completely true. It gets called as long as there are GST_MESSAGE_BUFFERING messages (there are a lot of updateBufferingStatus() calls between 0:00:51.206954836 and 0:00:51.498112047 (last time before finishing seek). The problem is that in all those cases any m_isBuffering true --> false transition is reverted and there's no spureous GST_MESSAGE_BUFFERING after the finished seek to unlock the blockage.

Now I'm trying to see how your solution would actually solve this problem, considering the 100% reaching that happens at 0:00:51.498112047. Thinking aloud here:

Let's suppose that 0:00:51.498112047 (the one that ultimately causes the seek "hang") is the case to be fixed, and let's also suppose that any other 100% level reach before that hasn't happened in these logs (for the sake of simplicity). In that case, we can assume that m_wasBuffering is true and m_isBuffering will now be false, triggering the true --> false transition and calling updateStates(). So far so good.

But... what will trigger more updateBufferingStatus() calls if not a new GST_MESSAGE_BUFFERING that will never come?

Maybe a better and more reliable solution to this "no more GST_MESSAGE_BUFFERING" messages before async change completion would be to reevaluate the buffering level in some way on async change completion, [here](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.46/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2103).

filipe-norte-red commented:
I may have over-simplified the execution steps. I didn't mean to say updateBufferingStatus() was never called again as it is called a few times due GST_MESSAGE_BUFFERING message as you say. I meant that at some point, and after a few calls to it happened, there will be no more GST_MESSAGE_BUFFERING messages and consequently no call to updateBufferingStatus(), which happens while the async transition is still in process.

> But... what will trigger more updateBufferingStatus() calls if not a new GST_MESSAGE_BUFFERING that will never come?

If we have already the transition for m_isBuffering from true --> false, then we would not need any more buffering messages as long as we don't reset that state as currently done [here](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/a710083012198719609db9c917afe6c03db24f87/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2889). If we keep resetting it, then something else is needed (which goes to your next point).

> Maybe a better and more reliable solution to this "no more GST_MESSAGE_BUFFERING" messages before async change completion would be to reevaluate the buffering level in some way on async change completion, [here](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/wpe-2.46/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2103).

In my first approach (which also worked for this use case without apparent side-effects during smoke testing), I did precisely the query to the buffering status and called updateBufferingStatus() (similar to the fillTimerFired() implementation). I did it before the updateSates() call [here](https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/a710083012198719609db9c917afe6c03db24f87/Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp#L2156), although updateBufferingStatus() was calling updateStates() already anyway.

I also agree that evaluating the buffering status again at a key point would be a more robust approach as it would not depend on some corner case execution sequence that may have not been identified yet. However, I had some concerns about the query operation being safe from within the message handler without the risk of causing blocks or dead-locks (despite done after state transition completed) although I could not pin point a specific case. There would also be an extra delay to the state transition completion handling, although probably irrelevant as it should be small enough (did not measure it) and acceptable in favor of a more robust solution.

I'm happy to change the solution if there is a more robust and safe way to do it.

filipe-norte-red committed:
Revert "Fix buffering handling in progressive playback"

This reverts commit 14bf12157a48e2b0ec47401567ba7e5db7fcce1d.

filipe-norte-red committed:
Fix buffering handling in progressive playback

When performing a seek operation during playback, the pipeline switches
to the PAUSED state and starts buffering data. As this transition is
async, we may reach the 100% buffer level before the state change
completes. This may cause updateBufferingStatus() to stop getting
called while the async transition is still in progress.
As m_isBuffering was getting set back to m_wasBuffering value
in updateStates() async state handling, once the state change
completed, the updateStates() might not detect that we are not
buffering anymore, as m_isBuffering is holding the wrong value.

filipe-norte-red commented:
I made these changes locally, but am seeing more 1-2 seconds video freezes after seek. I'll re-check this and update here the findings once I'm back

eocanha commented:
@filipe-norte-red Could you check if https://github.com/WebPlatformForEmbedded/WPEWebKit/commit/ac4f27bf217df457fe07c8946f9ab12d5563b80a (proposed by Volodymyr, from #1561) fixes the issue in your use case too?
